### PR TITLE
MO hybrid features update.   

### PR DESCRIPTION
**Pull request by ukmo-jianguo-li**
Updated atomic lines and hybrid MPI_INIT_THREAD

**Edit: The updates to MPI_INIT_THREAD have been moved to a separate PR**  #4 

This PR focusses on updates to the OMP directives in the `w3psmcmd` routine. This includes:
   - A bugfix where the `OMPH` switch was not specified correctly
   - Addition of `OMP ATOMIC` lines in preference of `OMP CRITICAL` sections; this included re-ordering some statements to allow atomic access.
   - A new `B4B` switch that ensures that certain calculations are bit-for-bit reproducible when run under OMP environment (workaround for floating point rounding error during sum).

The new B4B switch is intended to be used for regression testing where B4B reproducibility is required, but disabled for "operational" runs where the small (and negligible) amount of variability due to floating point rounding errors is not a concern.

### CODE REVIEW

**Comment by ukmo-ccbunney** [comment]
@ukmo-jianguo-li can you please run a regression test (`ww3_tp2.10` would be obvious choice) on your branch to ensure your changes don't have any unintended side effects ? This should be compared against the output of the same regression test run under the **develop** branch.

**Comment by ukmo-ccbunney** [comment]
> 
> 
> Updated atomic lines and hybrid MPI_INIT_THREAD

The MPI_INIT_THREAD changes have been moved to a new PR #4

**Comment by ukmo-ccbunney** [comment]
@CarstenHansen and @RobertoPadilla-NOAA - would you be willing to review this PR? Most of the changes are in the the w3psmcmd module; it might be easier to view the changes for that file via the separate commits (a lot of the changes were to fix an incorrectly specified switch).

**Comment by ukmo-ccbunney** [comment]
Results of regtests (ww3_tp2.10 and ww3_tp.216) attached.

Small differences in grid and point output due to reordering some code and rearranging some terms to allow atomic access.

Multiple runs of the new hybrid regtests (MPI_OMPH) show identical results using B4B switch.

[matrixCompSummary.zip](https://github.com/ukmo-waves/WW3/files/4318063/matrixCompSummary.zip)

**Comment by CarstenHansen** [comment]
Hi Chris and @ukmo-jianguo-li. I will try to review this PR. Although I have no practical experience with the SMC grids and have only sporadic knowledge about OpenMP, I will consult the Spatial Propagation section in the WW3 User Manual and the OpenMP 5.0 Specification on [https://www.openmp.org](https://www.openmp.org).

**Comment by ukmo-ccbunney** [comment]
Thanks @CarstenHansen. I have invited you as an external collaborator to our WW3 Repo - this will allow me to formally add you as a reviewer once you accept the invitation. I have also done the same for @RobertoPadilla-NOAA

**Comment by ukmo-ccbunney** [comment]
By the way - if you agree with the addition of the `B4B` switch, then we will add a commit to update the manual accordingly.

**Review by CarstenHansen — COMMENTED** [review]
Introduction of the B4B switch is a good idea as long as the side gradients are much larger than 1.e-6 m/s. I believe this is the case for the regression tests, and I believe the B4B switch will not be applied for a large-scale model.

File: model/ftn/w3psmcmd.ftn
Comment (CarstenHansen) [inline model/ftn/w3psmcmd.ftn:468]: Do you know that without the parentheses the first difference might be evaluated first and be stored in the memory of AFCN(M)? In that case, AFCN(M) would be updated twice, and I believe the OpenMP routines may miss one of the updates. The order of evaluation in lines 482 and 632 must also be changed.

EDIT: The parentheses are indeed required according to an [Intel® Fortran Compiler 19.1 Developer Guide and Reference](https://software.intel.com/en-us/fortran-compiler-developer-guide-and-reference-atomic). One of the rules that applies to an update-statement is:
"_x operator expr_ must be mathematically equivalent to _x operator (expr)_. This requirement is satisfied if the operators in expr have precedence greater than operator, or by using parentheses around expr or subexpressions of expr."

File: model/ftn/w3psmcmd.ftn
Comment (CarstenHansen) [inline model/ftn/w3psmcmd.ftn]: Change _x operator expr_  to _x operator (expr)_:             AFCN(N) = AFCN(N) + ( FUMD(i)*UCFL(N) - FUDIFX(i) )

File: model/ftn/w3psmcmd.ftn
Comment (CarstenHansen) [inline model/ftn/w3psmcmd.ftn]: Change _x operator expr_  to _x operator (expr)_: AFCN(M) = AFCN(M) + ( FUMD(i)*UCFL(M)*FMR - FUDIFX(i) )

**Review by CarstenHansen — CHANGES_REQUESTED** [review]
There are two more lines where I think _x operator expr_  should be changed to _x operator (expr)_. See my comments in commit e2cfc986019572f5e96acf580a7a83268c2b23d9

**Comment by ukmo-jianguo-li** [comment]
All,
It is no harm to add the parentheses for the calculations, at least on our supercomputer, if other compilers require.  I have done that in my test branch and tested with our UK 1.5-3km model. The results are identical to those without the parentheses.
Kind regards.
Jian-Guo

**Comment by ukmo-ccbunney** [comment]
@aliabdolali - would you be willing to review this PR?

**Comment by aliabdolali** [comment]
@ukmo-ccbunney Have you compared the entire regression test or just tp2.10 and tp2.16? If you believe it still do not affect the rest of the code, it is a necessity to have the entire suite comparison. I like the idea of B4B swtich, have you added the description in the manual for it? 
Happy Nowruz. 
AA

**Comment by ukmo-ccbunney** [comment]
Ali, Happy Nowruz too :) 

> 
> Have you compared the entire regression test or just tp2.10 and tp2.16?

I have just performed regtests for tp2.10 and tp2.16 as these are the only regtests that use the SMC module. I was assuming that the full regtest suite would be run at the point where I raise a PR to have our UKMO 'develop' branch merged with NOAA's. I can run a full suite if you prefer. This procedure is something that we might want to make clearer in our "review framework for trusted institutional forks" document. @ajhenrique - any comment on this?

EDIT: Actually - on re-reading the review framework document, I see that I should perform a full regression test matrix for each PR submitted to our repo. I will do this now.

>
> I like the idea of B4B swtich, have you added the description in the manual for it?

I have - I was just waiting for the reviews to agree that it was a good idea, then I intended to push the commit with the manual addition. Since you also agree it is worthwhile, I will push that change up now.

Cheers,
Chris.

**Comment by ukmo-ccbunney** [comment]
@aliabdolali and @CarstenHansen  - I've pushed a final commit that adds explicit handling of the `B4B`  switch in the `make_makefile.sh` and `w3_new` build scripts along with an update to the manual. Thanks both.

**Review by aliabdolali — APPROVED** [review]

**Review by CarstenHansen — APPROVED** [review]

**Comment by ukmo-ccbunney** [comment]
Full regtest results attached.

Summary: small differences to the SMC regtests only (ww3_tp2.10 and ww3_tp2.16); these are expected due to rearrangement of some terms to allow for ATOMIC access.

Max absolute difference in output Hs fields: 4.2143278e-05m and 5.9604645e-08m for ww3_tp2.10 and ww3_tp2.16 respectively.

[matrixCompSummary_openmp.zip](https://github.com/ukmo-waves/WW3/files/4402114/matrixCompSummary_openmp.zip)

**Comment by ukmo-jianguo-li** [comment]
Merged #3<https://github.com/ukmo-waves/WW3/pull/3> into develop.
Does this mean we can delete the fb_openmp branch now?   This is also a test of github recorded reply by email.

**Comment by ukmo-jianguo-li** [comment]
Well done!  I will clear my fb_branch copy from the desktop disk.  By the way, reply by email is tested fine as long as you tidy it up before hitting the send button.

**Comment by ukmo-ccbunney** [comment]
@ukmo-jianguo-li 
>Does this mean we can delete the fb_openmp branch 

Yes - you can! Remember, just deleting the repository from your local disk does not delete the branch from the GitHub repo. Feel free to clean up your local repo and I will delete the branch from our origin on GitHub. 

Edit: I will keep the branch alive here on GitHub until our develop branch has been merged with NOAA's authoritative repo...just in case.

**Pull request merged**
